lib/charms/layer/kubernetes_master.py

```diff
@@ -141,13 +149,170 @@ def get_cephfs_fsname():
+    In 1.19+, file-based authentication was deprecated in favor of webhook
     auth. Write out generic files that inform the user of this.
     """
     csv_file = Path(auth_file)
     csv_file.parent.mkdir(exist_ok=True)
+
+    csv_backup = Path('{}.{}'.format(csv_file, AUTH_BACKUP_EXT))
+    if csv_file.exists() and not csv_backup.exists():
+        csv_file.rename(csv_backup)
     with csv_file.open('w') as f:
-        f.write('# File-based authentication was deprecated in 1.19\n')
+        f.write('# File-based authentication was removed in Charmed Kubernetes 1.19\n')
+
+
+def migrate_auth_file(filename):
+    '''Create secrets or known tokens depending on what file is being migrated.'''
+    with open(str(filename), 'r') as f:
+        rows = list(csv.reader(f))
+
+    for row in rows:
+        try:
+            if row[0].startswith('#'):
+                continue
+            else:
+                if filename == AUTH_BASIC_FILE:
+                    create_known_token(*row)
+                elif filename == AUTH_TOKENS_FILE:
+                    create_secret(*row)
+                else:
+                    # log and return if we don't recognize the auth file
+                    hookenv.log('Unknown auth file: {}'.format(filename))
+                    return False
+        except IndexError:
+            pass
+    deprecate_auth_file(filename)
+    return True
+
+
+def token_generator(length=32):
+    '''Generate a random token for use in account tokens.
+
+    param: length - the length of the token to generate
+    '''
+    alpha = string.ascii_letters + string.digits
+    token = ''.join(random.SystemRandom().choice(alpha) for _ in range(length))
+    return token
+
+
+def create_known_token(token, username, user, groups=None):
+    known_tokens = Path(AUTH_TOKENS_FILE)
+    known_tokens.parent.mkdir(exist_ok=True)
+    csv_fields = ['token', 'username', 'user', 'groups']
+
+    try:
+        with known_tokens.open('r') as f:
+            tokens_by_user = {r['user']: r for r in csv.DictReader(f, csv_fields)}
+    except FileNotFoundError:
+        tokens_by_user = {}
+    tokens_by_username = {r['username']: r for r in tokens_by_user.values()}
+
+    if user in tokens_by_user:
+        record = tokens_by_user[user]
+    elif username in tokens_by_username:
+        record = tokens_by_username[username]
+    else:
+        record = tokens_by_user[user] = {}
+    record.update({
+        'token': token,
+        'username': username,
+        'user': user,
+        'groups': groups,
+    })
+
+    if not record['groups']:
+        del record['groups']
+
+    with known_tokens.open('w') as f:
+        csv.DictWriter(f, csv_fields, lineterminator='\n').writerows(
+            tokens_by_user.values())
+
+
+def create_secret(token, username, user, groups=None, ns='auth-webhook'):
+    sani_name = re.sub('[^0-9a-zA-Z]+', '-', user)
+    secret_id = '{}-secret'.format(sani_name)
+    delete_secret(secret_id, ns=ns)
```

Comment:
This was happening because we always called `delete_secret` early in `create_secret` -- we did that because we were using `kubectl create` which would have failed if the user already existed... Now that we're using `kubectl apply`, we can just stomp over existing secrets willy nilly.

Before:
```bash
  results:
    Stderr: |
      Error from server (NotFound): secrets "t1" not found
    Stdout: |
      Cluster "juju-cluster" set.
      Property "users" unset.
      User "t1" set.
      Context "juju-context" created.
      Switched to context "juju-context".
    kubeconfig: juju scp kubernetes-master/0:/home/ubuntu/t1-kubeconfig .
    msg: User "t1" created.
    users: admin, system:kube-controller-manager, system:kube-proxy, system:node:ip-172-31-0-211,
      system:node:ip-172-31-11-121, system:node:ip-172-31-22-158, system:kube-scheduler,
      system:monitoring, t1
  status: completed
```

Now:
```bash
  results:
    Stdout: |
      Cluster "juju-cluster" set.
      Property "users" unset.
      User "t2" set.
      Context "juju-context" created.
      Switched to context "juju-context".
    kubeconfig: juju scp kubernetes-master/0:/home/ubuntu/t2-kubeconfig .
    msg: User "t2" created.
    users: admin, system:kube-controller-manager, system:kube-proxy, system:node:ip-172-31-0-211,
      system:node:ip-172-31-11-121, system:node:ip-172-31-22-158, system:kube-scheduler,
      system:monitoring, t1, t2
  status: completed

```

Given this change, I think it would be useful to leave stderr in-tact if delete fails.